### DISCUSSION TITLE
Hydra w/external IdP (Azure AD)

### DISCUSSION THREAD

**Discussion by jfarleyx** [ory/hydra#2406] [Q&A] [answered]
I've got Hydra running in K8S and I can perform the client credential flow. That process is pretty straight forward. However, I need to set up and test using Hydra with an external identity provider via OIDC for the auth code flow. In this case, the IdP would be Microsoft Azure Active Directory. 

I'm a little confused on exactly how to configure Hydra to work with Azure AD as an OIDC IdP (or any 3rd party IdP). We are not in control of the identity provider, we just need our app & Hydra to work with that provider for user authentication via OIDC. Has anyone here configured Hydra to work with an external identity provider like Azure AD (or even Okta/Auth0)?

This is the flow I'm trying to accomplish:
1. User hits a web page I control. The page would offer a "Login w/ Azure AD" button, which should direct user to Azure controlled login page. I have examples from MS regarding how to use their OIDC SDK to authenticate users in Azure AD. That part is clear. 
2. User authenticates, letting our app know it was successful/unsuccessful. 
3. If successful authentication, we get access/refresh token for authorization from our Hydra instance. 

Questions:
Is Hydra able to accommodate this or am I misusing/misunderstanding Hydra?
Does Azure AD communicate with Hydra in this flow and if so, how exactly? What do I configure in Hydra to accommodate that communication. 
The Hydra login/consent/logout urls - I'm confused as to what to put here. In this use case, would I provide URL's to Azure AD login or are these URL's for an app we build? I suspect the latter. 

Basically, I'm not clear on how, if at all, the IdP and Hydra work together and how to configure Hydra to accommodate that relationship. Perhaps they don't communicate and my app simply manages them separately. If they are separate, can I simply leave the consent URL parameter blank for Hydra, as the consent UI is managed by Azure AD?  

If anyone has done this and could help me better understand how to set this up I would be extremely grateful.

**Comment by vinckr** [comment, marked as answer]
Could you use Ory Kratos instead? You can login with any OIDC provider, we also have a guide for [Azure AD ](https://www.ory.sh/kratos/docs/next/guides/sign-in-with-github-google-facebook-linkedin/#microsoft).
Maybe the [login flow for API clients](https://www.ory.sh/kratos/docs/self-service/flows/user-login#login-for-api-clients)?
It sounds to me as you are on the OAuth2/OIDC client side, rather than looking to become a OIDC server yourself.

**Reply by jfarleyx** [reply]
Thank you for responding! 

We are a hybrid environment, I guess. We have client apps & API's to protect, but we integrate with 3rd parties for user identification. We don't want to manage usernames & passwords and our customers generally have some sort of OIDC compatible IdP system (Azure AD, Okta, Auth0). So, my intent was to use Hydra for authorization and integrate with 3rd party OIDC compatible IdP for authentication. I just wasn't clear if there was any functionality in Hydra to facilitate the OIDC process and if there was, exactly how to configure it. 

So, based on your response I suspect that there isn't any communication that occurs between Hydra and an OIDC compatible IdP source. Instead, we would have to use Kratos to facilitate the OIDC process or create our own app?

**Reply by vinckr** [reply]
No sorry, I just misunderstood your environment.
I think you can use Hydra as you intended yes, if you look at this basic setup. 
<img width="1194" alt="ory-ecosystem" src="https://user-images.githubusercontent.com/55394283/111302513-c9bdd180-8653-11eb-8737-88a2de990374.png">
So you can switch out Kratos in this instance for the external IdP (Azure ect.), and then use the [OIDC flow](https://www.ory.sh/hydra/docs/concepts/openid-connect-oidc) to get OAuth tokens for those credentials. 
You need some login interface for Hydra, you can check out this reference implementation: https://github.com/ory/hydra-login-consent-node . (not sure maybe you can leave it out too.)

I think you would want to use the [Authorization Code Grant](https://oauth.net/2/grant-types/authorization-code/) rather than client credentials. 
As for how exactly the setup would work, I dont know. But would be happy to work with you on that and maybe we can add to the documentation. 
I will put something together how I imagine the flow.

**Reply by jfarleyx** [reply]
Sorry for the confusion and I really appreciate your assistance. I'll try to clarify my process a bit more. Also, I'm obviously not an OAuth expert, so please excuse any misuse of terms or misunderstanding of OAuth processes. :)

The plan is to use the auth code flow w/pkce in our client app, but direct the user to their IdP's login (in this case, Azure AD) for authentication. What I wasn't sure of is if Hydra could be configured to use the IdP for authn (e.g. set hydra login url to IdP login page) or if my client app had to facilitate the authn process on it's own and only interact with Hydra for authz. 

Ideally, my client app would only communicate with Hydra, requesting an auth code first, and Hydra would detect that the user has not been authenticated and would automatically redirect the browser to the IdP's login page. Then, when the user successfully logged in the IdP page would communicate the id token back to Hydra and Hydra could then communicate that back to client app. Then, the client app could continue to the authz process with Hydra. In other words, the client app communicates only with Hydra and Hydra would help to facilitate both authn & authz processes. 

Hope that helps clarify what I'm looking for. If the scenario above can be accomplished with Hydra, do I simply set the login url to the IdP's login URL and Hydra will know what info to pass to that login page to support the authentication process (for example, the client id)?

**Reply by vinckr** [reply]
Dont worry I am still learning with OAuth2 as well, and there is many different use cases :). The terms can be very easily mixed up (the different sessions for example), so I am trying to be as clear as possible, but do let me know if there is something confusing.

>What I wasn't sure of is if Hydra could be configured to use the IdP for authn (e.g. set hydra login url to IdP login page) 

Hydra does not implement a login interface so yes. 

>If the scenario above can be accomplished with Hydra 

I dont see why not, have a look at the login [flow diagram](https://www.ory.sh/hydra/docs/concepts/login). 

You would have to skip the consent stage I guess? But if you substitute the Login Endpoint with the Idp Login Endpoint, it should work.
There is probably some things would need to configure I am missing here, but for start I dont see why it shouldnt work.

**Reply by jfarleyx** [reply]
Thanks, Vincent. I'll review the flow diagram and work on mocking a test in our environment. I'll follow up here with results once I've concluded the test.

**Reply by vinckr** [reply]
Hey ,
Just checking in if you had a chance to test the setup @jfarleyx .

**Reply by jfarleyx** [reply]
Hi @vinckr I had to temporarily switch to another project and put this on hold. I should return to this project in the next few weeks and will continue with my test case. I'll follow up with details soon. Thanks for following up.

**Reply by michaelwagler** [reply]
I'm also interested in this. I'm not sure if an external IDP could replace the login endpoint, because from the Hydra docs it looks like the login endpoint has to do some specific things, like make calls back to hydra with the contents of the hydra_challenge as payload, which the external IDP probably can't do.

What I'm wondering is if you could implement /login in your own custom app, and then have that app delegate the actual login to a third-party IDP. However it might be a pretty poor user experience, since it sounds like Hydra requires the /login endpoint itself to return a login form, whereas for the flow I'd like to do, the user would get redirected to the third-party IDP from the /login endpoint, and the third-party IDP itself would have its own login form.

**Reply by tn185075** [reply]
Can you provide more details into such kind of setup @vinckr? We want to replicate it locally with Hydra and an external IDP.

**Reply by atreya2011** [reply]
@vinckr @tn185075
Hey, in this discussion I am talking about integrating Hydra with Kratos.

https://github.com/ory/hydra/discussions/2873

I have 3 separate branches, demonstrating, Kratos only, Kratos with Microsoft OIDC and Kratos with Hydra.   If my understanding is correct, what is required is a Kratos + Hydra integration with Kratos rendering a login form showing a `Sign in with Microsoft` button?

If that is the case, this can be achieved easily by integrating the Kratos + Hydra branch with the branch demonstrating Kratos + Microsoft OIDC.

Let me know if my understanding is correct 🙇🏼‍♂️

**Reply by tn185075** [reply]
> @vinckr @tn185075 Hey, in this discussion I am talking about integrating Hydra with Kratos.
> 
> #2873
> 
> I have 3 separate branches, demonstrating, Kratos only, Kratos with Microsoft OIDC and Kratos with Hydra. If my understanding is correct, what is required is a Kratos + Hydra integration with Kratos rendering a login form showing a `Sign in with Microsoft` button?
> 
> If that is the case, this can be achieved easily by integrating the Kratos + Hydra branch with the branch demonstrating Kratos + Microsoft OIDC.
> 
> Let me know if my understanding is correct 🙇🏼‍♂️


I wanted to integrate hydra with an external Identity provider, as you did with Kratos (with Microsoft).

**Reply by atreya2011** [reply]
> I wanted to integrate hydra with an external Identity provider, as you did with Kratos (with Microsoft).

@tn185075 My understanding is that as long as you can host said external Identity Provider on-premise, then it should be technically possible to delegate the login part of Hydra to said identity server. What I have done is integrate an external Identity provider with Kratos and delegate the login part of Hydra to Kratos.

**Reply by tn185075** [reply]
Ok! Does it work only for an on-premise IdP or, can we redirect to something on cloud? Can you share some steps in detail, for your hydra-kratos integration?

**Reply by atreya2011** [reply]
As vinckr pointed out, 
> It sounds to me as you are on the OAuth2/OIDC client side, rather than looking to become an OIDC server yourself.

If it is something that requires redirection to a provider-hosted on the cloud to which you have no control over, then a `kratos + oidc provider` solution is better suited rather than Hydra.

Hydra sends a login challenge to the identity provider and the identity provider should accept the login request from Hydra. Details can be found here: https://www.ory.sh/hydra/docs/guides/login#accepting-the-login-request

So if you have control over your cloud-hosted identity provider and can accept the login request from Hydra, you can integrate both of them.

**Reply by tn185075** [reply]
Yeah, a bit of both.

**Reply by tn185075** [reply]
will look at those resources. Thank you.

**Reply by atreya2011** [reply]
Sure, you can find the source code for my integration of Kratos with Hydra over here: https://github.com/atreya2011/go-kratos-test/blob/hydra-consent/main.go

**Reply by tn185075** [reply]
Thanks for sharing. Will look at it too. So, if we want a direct integration with Hydra, we need those handlers setup on the IDP. Am I correct?

**Reply by atreya2011** [reply]
That is correct, in the simplest form you would need 2 handlers:
- Accepting the Login Request
- Handling Consent

**Reply by tn185075** [reply]
@atreya2011 

`If that is the case, this can be achieved easily by integrating the Kratos + Hydra branch with the branch demonstrating Kratos + Microsoft OIDC.`

Did you achieve this?

**Reply by atreya2011** [reply]
@tn185075 
I didn't get the time to try out the implementation yet. You can technically do this by adding the Microsoft OIDC config values to the Kratos config. Basically, you merge the `kratos.yml` in the `microsof-oidc` branch with the `kratos.yml` present in the `hydra` branch.

**Reply by tn185075** [reply]
Any references on doing it without Kratos in the middle, and directly implementing it in the handlers of Hydra.

**Reply by atreya2011** [reply]
Hmmm, can you make sure that the Identity Provider can respond to the login challenge? You may need to make an intermediate endpoint for just accepting the login challenge and then redirecting to an external Identity Provider. Like what is mentioned over here:

https://github.com/ory/hydra/discussions/2406#discussioncomment-1241414

**Reply by atreya2011** [reply]
If all that is required is to integrate an external Identity Provider which is an `OAuth Server`, I think what is required is to actually implement an `OAuth Client`.
Hydra is an `OAuth Server` and not an `OAuth Client`

**Reply by tn185075** [reply]
yes we can have an intermediate endpoint, which accepts the login challenge, then how can we use the response from redirection to accept the login request is my question?

**Reply by tn185075** [reply]
So, we need an oauth client in between IDP and Hydra.

**Reply by atreya2011** [reply]
@tn185075 Please clarify if my understanding is correct. Do you want to host Hydra, which is an OAuth Server, that redirects to an external Identity Provider that is another OAuth Server? If you are looking to create OAuth Client (hosted by you) that directly connects to an external identity provider that is not hosted by you, then Kratos is more than enough.

**Reply by atreya2011** [reply]
Is this what you want to achieve?

![image](https://user-images.githubusercontent.com/44151328/145706281-64a9dca5-4313-4678-bd7d-62cd6d7acd3a.png)

**Reply by tn185075** [reply]
yeah, I want to redirect login requests from Hydra to external OAuth server (IDP). But don't want to use kratos here, wanna achieve it with Hydra.

**Reply by ondrejtomcik** [reply]
@tn185075 were you able to achieve this without Kratos? Directly Hydra <-> exIDP (e.g. Azure AD?)

Thanks

**Reply by tn185075** [reply]
yes @ondrejtomcik we need to handle the flow in login & consent handlers of Hydra.

**Reply by ondrejtomcik** [reply]
> yes @ondrejtomcik we need to handle the flow in login & consent handlers of Hydra.

So you wrote your own layer which uses hydra SDK and handles oidc. Right? Is it open-sourced?

**Reply by tn185075** [reply]
exactly. Sorry, it isn't open source.

**Reply by atreya2011** [reply]
@tn185075 Sorry for not getting back sooner. Glad to know you got it working without Kratos!

**Reply by astor4ever** [reply]
@tn185075 is there any chance to get some sample code to get better understand of that?

**Reply by tn185075** [reply]
I can try sending out the link to a gist here, if I find some time.

**Reply by astor4ever** [reply]
@tn185075 would be nice - also if You dont want to publish it always You can send it to me 😀

**Reply by astor4ever** [reply]
@tn185075 hey did you had time to prepare something? :)

**Reply by tn185075** [reply]
Will take some time tomorrow to publish. Hope that works. @astor4ever

**Comment by vinckr** [comment]
Hello @jfarleyx @tn185075 
The Ory Kratos and Ory Hydra integration just landed with Hydra 2.0, 
see the [release notes](https://github.com/ory/hydra/releases/tag/v2.0.0) for more details!
That will make this much easier!
